codegen/smithy-dafny-codegen/src/main/java/software/amazon/polymorph/smithyrust/generator/AbstractRustShimGenerator.java

```diff
@@ -0,0 +1,1153 @@
+              ","
+            )
+          )
+      )
+      .lineSeparated();
+  }
+
+  private TokenTree fromDafnyForMember(MemberShape member) {
+    Shape targetShape = model.expectShape(member.getTarget());
+    boolean isRequired = member.hasTrait(RequiredTrait.class);
+    // isRustOption is always true here because we're using .set_foo(...) fluent builder methods
+    // on the Rust side as opposed to just .foo(...), and those all take options
+    // even for required members.
+    return fromDafny(
+      targetShape,
+      "dafny_value." + member.getMemberName() + "()",
+      true,
+      !isRequired
+    );
+  }
+
+  /**
+   * @param isRustOption Whether the Rust type is an Option<...> or not.
+   *                     Rust's structures avoid Options when not strictly necessary depending on context.
+   * @param isDafnyOption Whether the Dafny type is an Option<...> or not.
+   *                      Dafny tends to be much more consistent about this.
+   */
+  // TODO: There is obviously a lot of duplication here that should be easy to clean up.
+  // TODO: Some cases do not handle all combinations of isRustOption and isDafnyOption.
+  private TokenTree fromDafny(
+    final Shape shape,
+    final String dafnyValue,
+    boolean isRustOption,
+    boolean isDafnyOption
+  ) {
+    return switch (shape.getType()) {
+      case STRING -> {
+        if (shape.hasTrait(EnumTrait.class)) {
+          var enumShapeName = toSnakeCase(shape.toShapeId().getName());
+          if (isDafnyOption) {
+            yield TokenTree.of(
+              """
+              match &**%s {
+                  crate::r#_Wrappers_Compile::Option::Some { value } => Some(
+                      crate::conversions::%s::from_dafny(value)
+                  ),
+                  _ => None,
+              }
+              """.formatted(dafnyValue, enumShapeName)
+            );
+          } else {
+            TokenTree result = TokenTree.of(
+              "crate::conversions::%s::from_dafny(%s)".formatted(
+                  enumShapeName,
+                  dafnyValue
+                )
+            );
+            if (isRustOption) {
+              result =
+                TokenTree.of(TokenTree.of("Some("), result, TokenTree.of(")"));
+            }
+            yield result;
+          }
+        } else {
+          if (isDafnyOption) {
+            yield TokenTree.of(
+              "crate::standard_library_conversions::ostring_from_dafny(%s.clone())".formatted(
+                  dafnyValue
+                )
+            );
+          } else {
+            TokenTree result = TokenTree.of(
+              "dafny_runtime::dafny_runtime_conversions::unicode_chars_false::dafny_string_to_string(%s)".formatted(
+                  dafnyValue
+                )
+            );
+            if (isRustOption) {
+              result =
+                TokenTree.of(TokenTree.of("Some("), result, TokenTree.of(")"));
+            }
```

Comment:
For the next PR, let's at least factor out this pattern, if not the other template for handling the combinations of isDafnyOption and isRustOption.